docs/stress-test.md

```diff
@@ -0,0 +1,124 @@
+# Hub Stress Testing
+
+This document gives an overview of the [hub-stress-test script](../scripts/hub-stress-test.py)
+and how it can be used.
+
+## Setup
+
+You will need two things to run the script, an admin token and a target hub API endpoint URL.
+
+The admin token can be provided to the script on the command line but it's recommended to create a
+file from which you can source and export the `JUPYTERHUB_API_TOKEN` environment variable.
+
+For the hub API endpoint URL, you can probably use the same value as the `JUPYTERHUB_API_URL`
+environment variable in your user notebooks, e.g. `https://myhub-testing.acme.com/hub/api`.
+
+Putting these together, you can have a script like the following to prepare your environment:
+
+```bash
+#!/bin/bash -e
+export JUPYTERHUB_API_TOKEN=abcdef123456
+export JUPYTERHUB_ENDPOINT=https://myhub-testing.acme.com/hub/api
+```
+
+## Scaling up
+
+By default the `hub-stress-test` script will scale up to 100 users and notebook servers (pods)
+in batches, wait for them to be "ready" and then stop and delete them.
+
+### Placeholders and user nodes
+
+The number of pods that can be created in any given run depends on the number of
+`user-placeholder` pods already in the cluster and the number of `user` nodes. The
+`user-placeholder` pods are pre-emptible pods which are part of a StatefulSet:
+
+```console
+$ kubectl get statefulset/user-placeholder -n jhub
+NAME               READY     AGE
+user-placeholder   300/300   118d
+```
+
+We normally have very few of these in our testing cluster but need to
+scale them up when doing stress testing otherwise the `hub-stress-test` script has to wait
+for the auto-scaler to add more nodes to the `user` worker pool. The number of available
+workers can be found like so:
+
+```console
+$ kubectl get nodes -l workerPurpose=users | grep -c "Ready\s"
+13
+```
+
+The number of `user` nodes needed for a scale test will depend on the resource requirements
+of the user notebook pods, reserved space on the nodes, other system pods running on the nodes,
+e.g. logging daemon, pod limits per node, etc.
+
+If there are not enough nodes available and the auto-scaler has to create them on the fly
+as the stress test is running, we can hit the [consecutive failure limit](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/363d0b7db5/jupyterhub/values.yaml#L17) which will cause the hub container to crash and restart.
+One way to avoid this is run the script with a `--count` that is not higher than 500 which
+gives time between runs for the auto-scaler to add more `user` nodes.
+
+As an example, on IBM Cloud there is a hard [pod limit](https://cloud.ibm.com/docs/containers?topic=containers-limitations)
```

Comment:
110 isn't really a hard limit. From the link:
>You can run 110 pods per worker node. If you have worker nodes with 11 CPU cores or more, you can support 10 pods per core, up to a limit of 250 pods per worker node. 

The 110 comes from the kubelet. It's the default value for `maxPods` in its [configuration](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/config/v1beta1/types.go#L468-L474) so it's likely that people will hit this issue with a lot of providers.

Reply:
I guess the 110 limit was specific for our worker node size since we have 8 CPUs on those so we're under that 250 pods per node threshold. Having said that, I think I also noticed that more global default of 110 limit per node (via some JupyterHub discourse thread) after writing this, so yeah this could be tweaked to just say there is by default 110 pods per node and drop the IBM Cloud mention.